File: iris-api/internal/api/catalog_cleanup_test.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"testing"
)

// The cleanup endpoint filters on two axes: which kinds are in scope, and which
// price is missing. These tests seed every kind × price-state combination and
// pin down exactly what each filter selects, end to end over HTTP.

// cleanupSeed is one seeded catalog item. Prices are raw JSON so a missing price
// is a real null rather than a zero.
type cleanupSeed struct {
	code     string
	kind     string
	purchase string
	sale     string
}

func cleanupSeeds() []cleanupSeed {
	seeds := make([]cleanupSeed, 0, 8)
	for _, kind := range []string{"service", "article"} {
		prefix := "S"
		if kind == "article" {
			prefix = "A"
		}
		seeds = append(seeds,
			cleanupSeed{code: prefix + "-none", kind: kind, purchase: "null", sale: "null"},
			cleanupSeed{code: prefix + "-purchase-only", kind: kind, purchase: "400", sale: "null"},
			cleanupSeed{code: prefix + "-sale-only", kind: kind, purchase: "null", sale: "1400"},
			cleanupSeed{code: prefix + "-both", kind: kind, purchase: "400", sale: "1400"},
		)
	}
	return seeds
}

// seedCleanupCatalog creates the full matrix through the public create endpoint,
// so the tests exercise the same write path the app uses.
func seedCleanupCatalog(t *testing.T, server *Server, adminToken string) {
	t.Helper()
	for _, seed := range cleanupSeeds() {
		body := fmt.Sprintf(
			`{"code":%q,"name":%q,"kind":%q,"unit":"kom","purchasePrice":%s,"salePrice":%s,"barcode":null,"taxGroup":null,"description":null,"isActive":true}`,
			seed.code, "Stavka "+seed.code, seed.kind, seed.purchase, seed.sale)
		if rec := roleRequest(t, server, adminToken, http.MethodPost, "/catalog-items", body); rec.Code != http.StatusCreated {
			t.Fatalf("create %s = %d (%s)", seed.code, rec.Code, rec.Body.String())
		}
	}
}

// catalogCodes lists the codes currently in the catalog, sorted.
func catalogCodes(t *testing.T, server *Server, adminToken string) []string {
	t.Helper()
	rec := roleRequest(t, server, adminToken, http.MethodGet, "/catalog-items", "")
	if rec.Code != http.StatusOK {
		t.Fatalf("list catalog items = %d (%s)", rec.Code, rec.Body.String())
	}
	var list struct {
		Items []struct {
			Code string `json:"code"`
		} `json:"items"`
		Total int `json:"total"`
	}
	if err := json.Unmarshal(rec.Body.Bytes(), &list); err != nil {
		t.Fatalf("decode catalog list: %v", err)
	}
	if list.Total != len(list.Items) {
		t.Fatalf("total = %d but got %d items", list.Total, len(list.Items))
	}
	codes := make([]string, 0, len(list.Items))
	for _, item := range list.Items {
		codes = append(codes, item.Code)
	}
	sort.Strings(codes)
	return codes
}

// survivors returns every seeded code except the deleted ones, sorted.
func survivors(deleted []string) []string {
	gone := map[string]bool{}
	for _, code := range deleted {
		gone[code] = true
	}
	remaining := make([]string, 0, len(cleanupSeeds()))
	for _, seed := range cleanupSeeds() {
		if !gone[seed.code] {
			remaining = append(remaining, seed.code)
		}
	}
	sort.Strings(remaining)
	return remaining
}

func sameCodes(got, want []string) bool {
	if len(got) != len(want) {
		return false
	}
	for i := range got {
		if got[i] != want[i] {
			return false
		}
	}
	return true
}

// TestCatalogCleanupFilterMatrix walks every (kinds, missing) combination the UI
// can produce and checks that the preview lists exactly the right items, the
// delete removes exactly those, and nothing else is touched.
func TestCatalogCleanupFilterMatrix(t *testing.T) {
	cases := []struct {
		name  string
		query string
		want  []string
	}{
		{
			name:  "services missing purchase price",
			query: "?kind=service&missing=purchase",
			want:  []string{"S-none", "S-sale-only"},
		},
		{
			name:  "services missing sale price",
			query: "?kind=service&missing=sale",
			want:  []string{"S-none", "S-purchase-only"},
		},
		{
			name:  "services missing both prices",
			query: "?kind=service&missing=both",
			want:  []string{"S-none"},
		},
		{
			name:  "articles missing purchase price",
			query: "?kind=article&missing=purchase",
			want:  []string{"A-none", "A-sale-only"},
		},
		{
			name:  "articles missing sale price",
			query: "?kind=article&missing=sale",
			want:  []string{"A-none", "A-purchase-only"},
		},
		{
			name:  "articles missing both prices",
			query: "?kind=article&missing=both",
			want:  []string{"A-none"},
		},
		{
			name:  "both kinds missing purchase price",
			query: "?kind=service&kind=article&missing=purchase",
			want:  []string{"A-none", "A-sale-only", "S-none", "S-sale-only"},
		},
		{
			name:  "both kinds missing sale price",
			query: "?kind=service&kind=article&missing=sale",
			want:  []string{"A-none", "A-purchase-only", "S-none", "S-purchase-only"},
		},
		{
			name:  "both kinds missing both prices",
			query: "?kind=service&kind=article&missing=both",
			want:  []string{"A-none", "S-none"},
		},
		{
			name:  "a repeated kind is not counted twice",
			query: "?kind=service&kind=service&missing=both",
			want:  []string{"S-none"},
		},
	}

	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			server, adminToken, _ := newServerWithRoles(t)
			seedCleanupCatalog(t, server, adminToken)

			// Preview: exactly the items that are about to be deleted.
			previewRec := roleRequest(t, server, adminToken, http.MethodGet, "/catalog-items/cleanup"+testCase.query, "")
			if previewRec.Code != http.StatusOK {
				t.Fatalf("preview = %d (%s)", previewRec.Code, previewRec.Body.String())
			}
			var preview struct {
				Items []struct {
					Code string `json:"code"`
				} `json:"items"`
				Total int `json:"total"`
			}
			if err := json.Unmarshal(previewRec.Body.Bytes(), &preview); err != nil {
				t.Fatalf("decode preview: %v", err)
			}
			previewCodes := make([]string, 0, len(preview.Items))
			for _, item := range preview.Items {
				previewCodes = append(previewCodes, item.Code)
			}
			sort.Strings(previewCodes)
			if !sameCodes(previewCodes, testCase.want) {
				t.Fatalf("preview = %v, want %v", previewCodes, testCase.want)
			}
			if preview.Total != len(testCase.want) {
				t.Fatalf("preview total = %d, want %d", preview.Total, len(testCase.want))
			}

			// Delete: the same set, and the count matches the preview.
			deleteRec := roleRequest(t, server, adminToken, http.MethodPost, "/catalog-items/cleanup"+testCase.query, "")
			if deleteRec.Code != http.StatusOK {
				t.Fatalf("cleanup = %d (%s)", deleteRec.Code, deleteRec.Body.String())
			}
			var result struct {
				Deleted int `json:"deleted"`
			}
			if err := json.Unmarshal(deleteRec.Body.Bytes(), &result); err != nil {
				t.Fatalf("decode cleanup response: %v", err)
			}
			if result.Deleted != len(testCase.want) {
				t.Fatalf("deleted = %d, want %d (%v)", result.Deleted, len(testCase.want), testCase.want)
			}
			if got, want := catalogCodes(t, server, adminToken), survivors(testCase.want); !sameCodes(got, want) {
				t.Fatalf("remaining catalog = %v, want %v", got, want)
			}

			// The preview is now empty and a repeat cleanup deletes nothing.
			repeatRec := roleRequest(t, server, adminToken, http.MethodGet, "/catalog-items/cleanup"+testCase.query, "")
			var repeat struct {
				Total int `json:"total"`
			}
			if err := json.Unmarshal(repeatRec.Body.Bytes(), &repeat); err != nil {
				t.Fatalf("decode repeat preview: %v", err)
			}
			if repeat.Total != 0 {
				t.Fatalf("preview after cleanup = %d, want 0", repeat.Total)
			}
		})
	}
}

// TestCatalogCleanupRejectsInvalidFilters proves a malformed or incomplete
// filter is refused outright rather than falling back to some wider default —
// and that nothing is deleted when it is.
func TestCatalogCleanupRejectsInvalidFilters(t *testing.T) {
	badQueries := []string{
		"",                                     // no parameters at all
		"?missing=both",                        // no kind
		"?kind=service",                        // no missing
		"?kind=everything&missing=both",        // unknown kind
		"?kind=service&missing=anything",       // unknown missing mode
		"?kind=service&kind=nope&missing=both", // one good kind, one bad
		"?kind=&missing=both",                  // blank kind
		"?kind=service&missing=",               // blank missing
	}

	for _, query := range badQueries {
		t.Run("query="+query, func(t *testing.T) {
			server, adminToken, _ := newServerWithRoles(t)
			seedCleanupCatalog(t, server, adminToken)
			before := catalogCodes(t, server, adminToken)

			previewRec := roleRequest(t, server, adminToken, http.MethodGet, "/catalog-items/cleanup"+query, "")
			if previewRec.Code != http.StatusBadRequest {
				t.Errorf("GET preview%s = %d, want %d", query, previewRec.Code, http.StatusBadRequest)
			}
			deleteRec := roleRequest(t, server, adminToken, http.MethodPost, "/catalog-items/cleanup"+query, "")
			if deleteRec.Code != http.StatusBadRequest {
				t.Errorf("POST cleanup%s = %d, want %d", query, deleteRec.Code, http.StatusBadRequest)
			}
			if after := catalogCodes(t, server, adminToken); !sameCodes(after, before) {
				t.Fatalf("catalog changed after a rejected cleanup: %v, want %v", after, before)
			}
		})
	}
}

// TestCatalogCleanupPreviewDoesNotDelete proves the preview is read-only: an
// admin can inspect what a cleanup would remove without removing it.
func TestCatalogCleanupPreviewDoesNotDelete(t *testing.T) {
	server, adminToken, _ := newServerWithRoles(t)
	seedCleanupCatalog(t, server, adminToken)
	before := catalogCodes(t, server, adminToken)

	for range 3 {
		if rec := roleRequest(t, server, adminToken, http.MethodGet,
			"/catalog-items/cleanup?kind=service&kind=article&missing=purchase", ""); rec.Code != http.StatusOK {
			t.Fatalf("preview = %d (%s)", rec.Code, rec.Body.String())
		}
	}
	if after := catalogCodes(t, server, adminToken); !sameCodes(after, before) {
		t.Fatalf("catalog changed after previews: %v, want %v", after, before)
	}
}

// TestCatalogCleanupRejectsNonAdminBeforeDeleting proves the operator gate holds
// on both verbs and that a rejected request leaves the catalog untouched.
func TestCatalogCleanupRejectsNonAdminBeforeDeleting(t *testing.T) {
	server, adminToken, userToken := newServerWithRoles(t)
	seedCleanupCatalog(t, server, adminToken)
	before := catalogCodes(t, server, adminToken)

	query := "?kind=service&kind=article&missing=both"
	for _, method := range []string{http.MethodGet, http.MethodPost} {
		if rec := roleRequest(t, server, userToken, method, "/catalog-items/cleanup"+query, ""); rec.Code != http.StatusForbidden {
			t.Errorf("%s as operator = %d, want %d", method, rec.Code, http.StatusForbidden)
		}
	}
	// Anonymous requests are rejected too.
	for _, method := range []string{http.MethodGet, http.MethodPost} {
		if rec := roleRequest(t, server, "", method, "/catalog-items/cleanup"+query, ""); rec.Code != http.StatusUnauthorized {
			t.Errorf("%s as anonymous = %d, want %d", method, rec.Code, http.StatusUnauthorized)
		}
	}
	if after := catalogCodes(t, server, adminToken); !sameCodes(after, before) {
		t.Fatalf("catalog changed after forbidden requests: %v, want %v", after, before)
	}
}

File: apps/web/src/lib/catalog-cleanup.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { createHttpApi } from './api-client'
import { createFixtureApi } from './fixture-api'
import type {
  CatalogCleanupFilter,
  CatalogCleanupMissing,
  CatalogItem,
  CatalogItemKind,
} from '@/types/catalog'

function response(body: unknown, init: ResponseInit = {}): Response {
  return new Response(JSON.stringify(body), {
    status: init.status ?? 200,
    headers: { 'Content-Type': 'application/json' },
  })
}

// The cleanup filter has two axes, so the tests below walk the full matrix of
// kinds × missing-price modes rather than spot-checking a few combinations.
interface MatrixCase {
  filter: CatalogCleanupFilter
  /** Codes of the seeded items this filter must select, sorted. */
  want: string[]
}

const services: CatalogItemKind[] = ['service']
const articles: CatalogItemKind[] = ['article']
const bothKinds: CatalogItemKind[] = ['service', 'article']

const matrix: MatrixCase[] = [
  { filter: { kinds: services, missing: 'purchase' }, want: ['S-none', 'S-sale-only'] },
  { filter: { kinds: services, missing: 'sale' }, want: ['S-none', 'S-purchase-only'] },
  { filter: { kinds: services, missing: 'both' }, want: ['S-none'] },
  { filter: { kinds: articles, missing: 'purchase' }, want: ['A-none', 'A-sale-only'] },
  { filter: { kinds: articles, missing: 'sale' }, want: ['A-none', 'A-purchase-only'] },
  { filter: { kinds: articles, missing: 'both' }, want: ['A-none'] },
  {
    filter: { kinds: bothKinds, missing: 'purchase' },
    want: ['A-none', 'A-sale-only', 'S-none', 'S-sale-only'],
  },
  {
    filter: { kinds: bothKinds, missing: 'sale' },
    want: ['A-none', 'A-purchase-only', 'S-none', 'S-purchase-only'],
  },
  { filter: { kinds: bothKinds, missing: 'both' }, want: ['A-none', 'S-none'] },
]

function caseName(filter: CatalogCleanupFilter): string {
  return `${filter.kinds.join('+')} missing ${filter.missing}`
}

describe('catalog cleanup HTTP encoding', () => {
  const emptyResult = { items: [], total: 0 }

  it.each(matrix.map(({ filter }) => [caseName(filter), filter] as const))(
    'encodes %s as repeated kind params plus missing',
    async (_name, filter) => {
      const fetchMock = vi.fn().mockResolvedValue(response(emptyResult))
      const api = createHttpApi('http://api.test', fetchMock)

      await api.previewCatalogCleanup(filter)

      const [requestUrl, init] = fetchMock.mock.calls[0]
      const url = new URL(requestUrl as string)
      expect(url.pathname).toBe('/catalog-items/cleanup')
      expect(url.searchParams.getAll('kind')).toEqual(filter.kinds)
      expect(url.searchParams.get('missing')).toBe(filter.missing)
      expect((init as RequestInit).method ?? 'GET').toBe('GET')
      expect((init as RequestInit).credentials).toBe('include')
    },
  )

  it('sends the delete to the same URL the preview used, as a POST', async () => {
    const filter: CatalogCleanupFilter = { kinds: bothKinds, missing: 'sale' }
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce(response(emptyResult))
      .mockResolvedValueOnce(response({ deleted: 0 }))
    const api = createHttpApi('http://api.test', fetchMock)

    await api.previewCatalogCleanup(filter)
    await api.cleanupCatalogItems(filter)

    const [previewUrl] = fetchMock.mock.calls[0]
    const [deleteUrl, deleteInit] = fetchMock.mock.calls[1]
    // Identical query strings: what the admin saw is what gets deleted.
    expect(deleteUrl).toBe(previewUrl)
    expect((deleteInit as RequestInit).method).toBe('POST')
    expect((deleteInit as RequestInit).credentials).toBe('include')
  })

  it('returns the deleted count from the API', async () => {
    const fetchMock = vi.fn().mockResolvedValue(response({ deleted: 7 }))
    const api = createHttpApi('http://api.test', fetchMock)

    await expect(
      api.cleanupCatalogItems({ kinds: services, missing: 'both' }),
    ).resolves.toEqual({ deleted: 7 })
  })

  it('surfaces a rejected filter as an error instead of reporting success', async () => {
    const fetchMock = vi
      .fn()
      .mockResolvedValue(response({ error: 'Neispravna vrsta stavke.' }, { status: 400 }))
    const api = createHttpApi('http://api.test', fetchMock)

    await expect(
      api.cleanupCatalogItems({ kinds: [], missing: 'both' }),
    ).rejects.toThrow('Neispravna vrsta stavke.')
  })

  it('tolerates a null items array in the preview response', async () => {
    const fetchMock = vi.fn().mockResolvedValue(response({ items: null, total: 0 }))
    const api = createHttpApi('http://api.test', fetchMock)

    await expect(
      api.previewCatalogCleanup({ kinds: services, missing: 'both' }),
    ).resolves.toEqual({ items: [], total: 0 })
  })
})

// The fixture API is the dev-mode data source; it has to agree with the server
// on exactly which items a filter selects, or fixtures mode would mislead.
describe('catalog cleanup against the fixture API', () => {
  type FixtureApi = ReturnType<typeof createFixtureApi>

  const seeds: { code: string; kind: CatalogItemKind; purchase: number | null; sale: number | null }[] = [
    { code: 'S-none', kind: 'service', purchase: null, sale: null },
    { code: 'S-purchase-only', kind: 'service', purchase: 400, sale: null },
    { code: 'S-sale-only', kind: 'service', purchase: null, sale: 1400 },
    { code: 'S-both', kind: 'service', purchase: 400, sale: 1400 },
    { code: 'A-none', kind: 'article', purchase: null, sale: null },
    { code: 'A-purchase-only', kind: 'article', purchase: 100, sale: null },
    { code: 'A-sale-only', kind: 'article', purchase: null, sale: 250 },
    { code: 'A-both', kind: 'article', purchase: 100, sale: 250 },
  ]

  let api: FixtureApi

  async function catalogCodes(): Promise<string[]> {
    const { items } = await api.getCatalogItems()
    return items.map((item: CatalogItem) => item.code).sort()
  }

  beforeEach(async () => {
    api = createFixtureApi()
    // The fixture catalog is module state shared across instances: clear it so
    // the matrix below is the whole world.
    const existing = await api.getCatalogItems()
    for (const item of existing.items) {
      await api.deleteCatalogItem(item.id)
    }
    for (const seed of seeds) {
      await api.createCatalogItem({
        code: seed.code,
        name: `Stavka ${seed.code}`,
        kind: seed.kind,
        unit: 'kom',
        purchasePrice: seed.purchase,
        salePrice: seed.sale,
        barcode: null,
        taxGroup: null,
        description: null,
        isActive: true,
      })
    }
  })

  it.each(matrix.map(({ filter, want }) => [caseName(filter), filter, want] as const))(
    'previews and deletes exactly the items for %s',
    async (_name, filter, want) => {
      const preview = await api.previewCatalogCleanup(filter)
      expect(preview.items.map((item: CatalogItem) => item.code).sort()).toEqual(want)
      expect(preview.total).toBe(want.length)

      const { deleted } = await api.cleanupCatalogItems(filter)
      expect(deleted).toBe(want.length)

      const remaining = seeds
        .map((seed) => seed.code)
        .filter((code) => !want.includes(code))
        .sort()
      expect(await catalogCodes()).toEqual(remaining)

      // Repeating the same cleanup is a no-op.
      expect((await api.cleanupCatalogItems(filter)).deleted).toBe(0)
    },
  )

  it('deletes nothing when no kind is selected', async () => {
    const before = await catalogCodes()
    for (const missing of ['purchase', 'sale', 'both'] as CatalogCleanupMissing[]) {
      const preview = await api.previewCatalogCleanup({ kinds: [], missing })
      expect(preview.items).toEqual([])
      expect((await api.cleanupCatalogItems({ kinds: [], missing })).deleted).toBe(0)
    }
    expect(await catalogCodes()).toEqual(before)
  })

  it('leaves the catalog untouched when only previewing', async () => {
    const before = await catalogCodes()
    await api.previewCatalogCleanup({ kinds: bothKinds, missing: 'purchase' })
    await api.previewCatalogCleanup({ kinds: bothKinds, missing: 'sale' })
    expect(await catalogCodes()).toEqual(before)
  })

  it('never deletes an item that has the price being filtered on', async () => {
    await api.cleanupCatalogItems({ kinds: bothKinds, missing: 'purchase' })
    const { items } = await api.getCatalogItems()
    for (const item of items) {
      expect(item.purchasePrice).not.toBeNull()
    }
  })
})
